data/basics/teal.go

```diff
@@ -0,0 +1,312 @@
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package basics
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/algorand/go-algorand/config"
+)
+
+// DeltaAction is an enum of actions that may be performed when applying a
+// delta to a TEAL key/value store
+type DeltaAction uint64
+
+const (
+	// SetUintAction indicates that a Uint should be stored at a key
+	SetUintAction DeltaAction = 1
+
+	// SetBytesAction indicates that a TEAL byte slice should be stored at a key
+	SetBytesAction DeltaAction = 2
+
+	// DeleteAction indicates that the value for a particular key should be deleted
+	DeleteAction DeltaAction = 3
+)
+
+// ValueDelta links a DeltaAction with a value to be set
+type ValueDelta struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	Action DeltaAction `codec:"at"`
+	Bytes  string      `codec:"bs"`
+	Uint   uint64      `codec:"ui"`
+}
+
+// ToTealValue converts a ValueDelta into a TealValue if possible, and returns
+// ok = false with a value of 0 TealUint if the conversion is not possible.
+func (vd *ValueDelta) ToTealValue() (value TealValue, ok bool) {
+	switch vd.Action {
+	case SetBytesAction:
+		value.Type = TealBytesType
+		value.Bytes = vd.Bytes
+		ok = true
+	case SetUintAction:
+		value.Type = TealUintType
+		value.Uint = vd.Uint
+		ok = true
+	case DeleteAction:
+		value.Type = TealUintType
+		ok = false
+	default:
+		value.Type = TealUintType
+		ok = false
+	}
+	return value, ok
+}
+
+// StateDelta is a map from key/value store keys to ValueDeltas, indicating
+// what should happen for that key
+//msgp:allocbound StateDelta config.MaxStateDeltaKeys
+type StateDelta map[string]ValueDelta
+
+// Equal checks whether two StateDeltas are equal. We don't check for nilness
+// equality because an empty map will encode/decode as nil. So if our generated
+// map is empty but not nil, we want to equal a decoded nil off the wire.
+func (sd StateDelta) Equal(o StateDelta) bool {
+	// Lengths should be the same
+	if len(sd) != len(o) {
+		return false
+	}
+	// All keys and deltas should be the same
+	for k, v := range sd {
+		// Other StateDelta must contain key
+		ov, ok := o[k]
+		if !ok {
+			return false
+		}
+
+		// Other StateDelta must have same value for key
+		if ov != v {
+			return false
+		}
+	}
+	return true
+}
+
+// Valid checks whether the keys and values in a StateDelta conform to the
+// consensus parameters' maximum lengths
+func (sd StateDelta) Valid(proto *config.ConsensusParams) error {
+	for key, delta := range sd {
+		if len(key) > proto.MaxAppKeyLen || proto.MaxAppKeyLen == 0 {
```

Comment:
can we move the `proto.MaxAppKeyLen == 0` check outside the loop in case len(sd) > 0 ?
that would retain the logical behavior, but would conduct the test only once per function call.

Reply:
done -- (also checked that `len(sd) > 0` to make it completely equivalent)